clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -11451,7 +11451,11 @@ bool Sema::CheckDeductionGuideDeclarator(Declarator &D, QualType &R,
     bool MightInstantiateToSpecialization = false;
     if (auto RetTST =
             TSI->getTypeLoc().getAsAdjusted<TemplateSpecializationTypeLoc>()) {
-      TemplateName SpecifiedName = RetTST.getTypePtr()->getTemplateName();
+      const TemplateSpecializationType *TST = RetTST.getTypePtr();
+      while (TST && TST->isTypeAlias())
+        TST = TST->getAliasedType()->getAs<TemplateSpecializationType>();
+
+      TemplateName SpecifiedName = TST->getTemplateName();
```

Comment:
Do we need to assert on `TST` in the case that we exit the above loop because it is null?

Reply:
@antangelo thanks for the review. I've added assert or would an additional condition be better?

Reply:
If there is some test case that would trip the assert (i.e. some input for which the alias type is not a template specialization type), then we should handle it more gracefully. If it (conceivably) can't happen, then an assert is fine.

I thought about it for a bit and came up with this (somewhat contrived) invalid code that might fall into this path:
```c++
template<class T> struct A {};

struct B {};

template<typename T>
using X = B;

template<typename T>
A() -> X<T>;
```
https://godbolt.org/z/shzK9eccz

Reply:
@antangelo thanks. I've removed the assert in favor of a condition and added additional tests.